### ISSUE TITLE
User-initiated zoom events?

### ISSUE THREAD

**Issue by geleto** [Leaflet/Leaflet#2934]
Hello,
It seems that leaflet has no specific event, which is triggered when the user zooms the map, but not when the zoom level was changed programmatically.
Maybe dragstart should be triggered when pinch-zooming because this usually changes the map position too.
I use the leaflet-locatecontrol and I need to stop tracking the user movements when he drags or pinch-zooms the map.

**Comment by Grsmto** [comment]
While creating a custom layer I found myself having to hack the lib to have access to zoom events.
I needed to know to which zoom level the map was animating, zoom property is updated only on `zoomend` event, which is too late if you want to animate anything custom at the same time as the map animation.
If anyone is interested by my solution, what I did is firing a `zoom` event:
`this.fire('zoom', {zoom: zoom});` in `setZoom()` and `setZoomAround()` functions from `L.Map`.

Can be related to https://github.com/Leaflet/Leaflet/issues/3269

**Comment by dmitry** [comment]
I have the same situation, when I need to prevent from invoking `fitBounds` while user interacts with the map. Looks like it could be fixed while https://github.com/Leaflet/Leaflet/pull/3278, but it's already was merged. I believe it could be named something like `triggerSource` with values `user` and `program`, or just a boolean `triggeredByUser`.

**Comment by dmitry** [comment]
Additional information, but related to the google maps: http://stackoverflow.com/questions/10097647/how-to-listen-for-user-generated-zoom-in-google-maps

**Comment by dmitry** [comment]
Just for anyone, who stuck with the same issue, here are the approach I've used in the project:

```javascript
  onDragEnd() {
    this.onUserChange();
  },
  onZoomEnd() {
    if (this.isUser) {
      this.onUserChange();
    }
  },
  onMoveEnd() {
    this.isUser = true;
  },
```

bindings:

```javascript
    this.map.on('dragend', this.onDragEnd);
    this.map.on('zoomend', this.onZoomEnd);
    this.map.on('moveend', this.onMoveEnd);
```

Would be nice to replace `onMoveEnd` to `idle` (https://github.com/Leaflet/Leaflet/issues/3178), but for my case that's enough.

**Comment by bpinney** [comment]
Any update on when this might be added?

**Comment by bpinney** [comment]
#2267 and #2532 are open issues making the same request

**Comment by yohanboniface** [comment]
The fastest way of getting this into core is a pull request ;)
We are now working on 1.0.0-rc1 release, and we have frozen the features perimeter of the 1.0, so we will basically only work on bug fixes ourselves. But any pull request will be considered.
So, to answer your question, unless someone provide a pull request, this will not be in Leaflet before 1.1 or later.

**Comment by Grsmto** [comment]
Didn't do any PR back in time as I saw you were rebuilding the whole things with 1.0.
Gonna have a look to that to implement the fix I did above.

**Comment by bpinney** [comment]
Yes @yohanboniface, doing stuff is always the fastest way to get stuff done :)
I would like some direction as to which solution is best before I take on implementation as the issue doesn't look very simple.  I noticed an event muting option for programmatic changes to map state via the API was suggested in https://github.com/Leaflet/Leaflet/issues/3088, this may be a solution as well.

**Comment by frankhommers** [comment]
It would be awesome if I could pass arbitrary data on the zoom function in the options, and get access to those options in the zoomstart and zoomend events.

This way I you could detect it yourself because you've passed something ;-) And also it would open a lot more possibilities.

@dmitry Did you find a better way in the mean time?

**Comment by mbrucher** [comment]
How can we make it better then?
